packages/@aws-cdk/cdk/lib/core/tag-manager.ts

```diff
@@ -151,3 +157,64 @@ export class TagManager {
     return true;
   }
 }
+
+class StandardFormatter implements ITagFormatter {
+  public renderTags(tags: ITag[], propertyTags: any): any {
+    const cfnTags: CfnTag[] = [];
+    for (const tag of tags) {
+      cfnTags.push({key: tag.key, value: tag.value});
+    }
+    const finalTags = mergeTags(propertyTags || [], cfnTags);
+    return finalTags.length === 0 ? undefined : finalTags;
+  }
+}
+
+interface AsgTag {
+  key: string;
+  value: string;
+  propagateAtLaunch: boolean;
+}
+
+class AsgFormatter implements ITagFormatter {
+  public renderTags(tags: ITag[], propertyTags: any): any {
+    const cfnTags: AsgTag[] = [];
+    for (const tag of tags) {
+      cfnTags.push({key: tag.key,
+        value: tag.value,
+        propagateAtLaunch: tag.props.applyToLaunchedInstances !== false});
+    }
+    const finalTags = mergeTags(propertyTags || [], cfnTags);
+    return finalTags.length === 0 ? undefined : finalTags;
+  }
+}
+
+class MapFormatter implements ITagFormatter {
+  public renderTags(tags: ITag[], propertyTags: any): any {
+    const cfnTags: {[key: string]: string} = {};
+    for (const tag of tags) {
+      cfnTags[tag.key] = tag.value;
+    }
+    const finalTags = mergeTags(propertyTags || {}, cfnTags);
+    return Object.keys(finalTags).length === 0 ? undefined : finalTags;
+  }
+}
+
+class NoFormat implements ITagFormatter {
+  public renderTags(_tags: ITag[], _propertyTags: any): any {
+    return undefined;
+  }
+}
+
+function mergeTags(target: any, source: any): any {
```

Comment:
I don't particularly like that this thing can handle both arrays and objects, because we lose type information this way. It should just handle arrays, and the `MapFormatter` should just do the object assignment directly.

Reply:
Fair - I did this so I could later flip the precedence, but moved it back as you requested.